Adding toolbar button actions to the menubar.

jaeschrich opened the pull request:
I added most of the toolbarbutton actions to the menubar. Right now, the menu items are copied-and-pasted from the toolbarbutton popup menus. I can refactor this to avoid duplication in one of two ways: (1) by creating an XBL binding that encapsulates the shared menu items, or (2) by creating a JavaScript function that sets up both the menubar popup and the toolbarbutton popup identically. I'm not sure which method of encapsulating the shared functionality is more idiomatic for the codebase, or which method will be easier to port away from XUL later.

The only toolbarbutton whose functionality I didn't add to the menubar is the sync button, because I'm not sure how best to expose that functionality via a menu. Right now the information about the sync status is displayed via a popup and there's a cancel button which is dynamically added to the toolbar to cancel a sync in progress. I don't think that adapting this behavior directly to the menubar makes sense. I could create a popup window for the menubar sync functionality, but that seemed beyond the scope of an individual pull request.

AbeJellinek committed:
Fix Scaffold load error when fontSize pref is set

AbeJellinek committed:
Scaffold templates: Pull URL from doc

Link URLs could lead to redirects.

dstillman committed:
Try to open external PDF reader to page number for "Show on Page"

PDF-handling code from `ZoteroPane.viewAttachment()` predates
`Zotero.OpenPDF` and didn't handle page numbers for external PDF
readers, so when called from "Show on Page" in a note (via
`ZoteroPane.viewPDF()`) it just launched the file itself in the
specified handler or via the OS.

For now, change to use `Zotero.OpenPDF.openToPage()` when there's a page
number. Later we should use `Zotero.OpenPDF` for everything, but that
code is more complicated than just using `launchFile()`, so we'll need
to do some testing (particularly on Linux) to see what's best.

(Note that "Show on Page" with an external reader doesn't make a ton of
sense, since you can't actually see the annotation, but opening the
right page is better than nothing.)

Fixes #2612

dstillman committed:
Update translators, styles, and CSL locales

dstillman committed:
Update locales from Transifex

mrtcode committed:
Avoid error interrupting annotation extraction from multiple items

dstillman committed:
Update note-editor submodule

dstillman committed:
Update version

dstillman committed:
Feed import: Don't fail on OPML entry with no `title` or `text`

https://forums.zotero.org/discussion/96841/impossible-dimporter-ompl-rss

dstillman committed:
Add prefs for timeouts for downloading PDF via browser

1 second isn't enough for some users, so try 1.5, and allow for easier
debugging via a pref.

https://forums.zotero.org/discussion/96431/i-have-access-to-pdf-but-zotero-cannot-automatically-add-it
https://forums.zotero.org/discussion/97277/unable-to-download-papers-using-zotero-chrome-extension

dstillman committed:
Update note-editor submodule

dstillman committed:
citeproc-js: Add space after left margin in plain-text mode

This adds a space after the the number in some numeric styles ("[1]" in
IEEE style, "1." in Nature) when pasting bibliography entries into a
plain-text editor.

Some styles (ACS, AMA, Vancouver) will need to be updated to remove
their hard-coded spaces to avoid double spaces after this change.

Patch from @zepinglee

Fixes #2633

dstillman committed:
Update locales from Transifex

dstillman committed:
Update global schema

Includes the Preprint archiveID/number mapping for #2481

dstillman committed:
Fix support tests after 28ed3e34b

dstillman committed:
Fix detection of some emoji set as colored tags

Emoji such as ⭐️ that began as text-based Unicode characters and depend
on Variation Selector-16 to display the emoji presentation weren't being
properly detected.

mrtcode committed:
Fix library tab tooltip always staying on top

adomasven committed:
Fix cookieSandbox not overriding user agent if no cookies for domain.

Potential fix for #2635

adomasven committed:
Improve downloadPDFViaBrowser handlers and logging

adomasven committed:
Fix a MIMETypeHandler.removeHandler() bug

adomasven committed:
Log full hidden browser count after its creation

mrtcode committed:
Add support for PDF page deletion and rotation (#2595)

* Add support for PDF page deletion and rotation
Fixes #2561

dstillman committed:
Update locales from Transifex and merge new English strings

AbeJellinek committed:
Fix sorting by attachment (#2586)

dstillman committed:
Update styles submodule

dstillman committed:
Fix Emoji detection

And move to Zotero.Utilities.Internal.isOnlyEmoji()

Fixes #2643

dstillman committed:
Properly strip ZWJ in `isOnlyEmoji()`

dstillman committed:
Fix parsing of Chinese and other non-ASCII month names

https://forums.zotero.org/discussion/97683/bug-locale-issues-concerning-zotero-and-zotero-connector

dstillman committed:
Use String.fromCharCode() to load Chinese characters in strToDate() test

Follow-up to dc045250a6

AbeJellinek committed:
Fix error if New Collection prompt is canceled

adomasven committed:
Fix force-click during drag sometimes causing collections to switch

Closes #2652

adomasven committed:
Fix alt-arrow not working on macOS when editing collection names.

Fixes #2651

adomasven committed:
Limit the force-click during drag fix to macOS (4fb0661b)

mrtcode committed:
Update pdf-reader and note-editor, use specific prebuilds, tweak code

dstillman committed:
Update translators

dstillman committed:
Update locales from Transifex

dstillman committed:
Update utilities submodule and date-parsing tests

Includes strToDate() fixes

dstillman committed:
Update version

mrtcode committed:
Implement smart notes switching in contextPane (#2671)

Fixes #2650

dstillman committed:
Update pdf-reader submodule

mrtcode committed:
Show parent item title for child notes in context pane
Fixes #2681

mrtcode committed:
Adjust page rotation labels in PDF reader
Fixes #2669

dstillman committed:
Update global schema and utilities for CSL 1.0.2

mrtcode committed:
Fix disabled "Add to Note" when a child note is open
Fixes #2687

tnajdek committed:
Don't try to remove temporary dir after import

Previously we had a structure that also created folders in temp dir,
that has since changed but code in cleanup phase to remove containing
dir has been accidentally left unremoved.

tnajdek committed:
Mendeley import: detect win close, better feedback

* Detect wizard cancel/close and interrupt import. This will still
  have to wait for current fetch (file or metadata) to complete but will
  then advance to the cleanup stage
* Advance progress bar during metadata fetch
* Add some extra logging

tnajdek committed:
Fix a bug in regex extracting fields to "extra"

Because regex is built using a template string, \s* is actually escaped
into s*, i.e. literal "s" appearing 0 or more times. In most cases this
would mean that output can have spacing slightly off. In extreme case,
when identifier starts with letter "s", this could this could lead to
identifier being stored incorrectly.

Also adjusted tests to be more strict and mock data to cover this case.

tnajdek committed:
Lint code related to Mendeley importer

tnajdek committed:
Improve log output for multiple Mendeley imports

Helps with debugging cases where new import is started while previous
one is still being cancelled

dstillman committed:
Validate fields in ItemTree::getSortFields()

To avoid startup hang if a plugin does something bad:

https://forums.zotero.org/discussion/comment/411843/#Comment_411843

Fixes #2692

dstillman committed:
Fix test failure after a8ed30ce8

We added a `toggleSort()` call on the title, so then when called again
it reverses the direction. To avoid, sort by creator and then title in
the second test.

dstillman committed:
Add CSL 1.0.2 locators and switch to CSL locales for localization

And move Page to middle of list, while still selecting by default

https://forums.zotero.org/discussion/comment/396602/#Comment_396602

dstillman committed:
Update locales from Transifex

dstillman committed:
Update pdf-reader and pdf-worker submodules

AbeJellinek committed:
`recursiveCollections`: Remove from all subcollections on delete (#2689)

dstillman committed:
Merge pull request #2693 from tnajdek/mendeley-import-improve-feedback

Multiple improvements to Mendeley importer

dstillman committed:
Tweak confirmation message for recursive collection-item removal

I guess the "its" is technically a little clearer, but it's a bit
awkward, and I think it's reasonably implied.

dstillman committed:
Merge new English strings

dstillman committed:
Add localized string for `timestamp` CSL locator

Not included in the CSL locales:

https://forums.zotero.org/discussion/comment/412060/#Comment_412060

dstillman committed:
Merge new English strings

dstillman committed:
CookieSandbox: Fix error setting request cookies

Regression from 6af3b0276a, showing up as "trackedBy is undefined" error

https://forums.zotero.org/discussion/98322/no-snapshot-or-pdf-autosaved-using-zotero-connector-on-safari

dstillman committed:
Increase `downloadPDFViaBrowser.onLoadTimeout` to 3000

Apparently necessary for some people to get PDFs on ScienceDirect. We
should come up with a better way of detecting the page load if we can,
so that we can avoid this long of a timeout for people who just will
never have access to ScienceDirect PDFs.

https://forums.zotero.org/discussion/comment/411056/#Comment_411056

Addresses #2698

dstillman committed:
Fix sorting by Item Type and feed item itemID

Regression from a8ed30ce803

dstillman committed:
Fix sorting of attachments and notes columns

Regression from a8ed30ce803

dstillman committed:
Sort empty rows last for Date/Year columns when descending

Date columns start in descending order, so put empty rows at end

Also remove a seemingly unnecessary exception for Title when not the
primary sort and just always sort it with empty rows first

dstillman committed:
Revert "Validate fields in ItemTree::getSortFields()"

This reverts commit a8ed30ce803a9c5b3b5c04d50f6bd6b94faeee86 and related
commits.

We'll address breakage from invalid sort fields another way, without
inconveniencing plugin authors.

https://groups.google.com/g/zotero-dev/c/kc0-C6-SA74/m/bhHniGceAQAJ

dstillman committed:
Better fix for errors from invalid sort fields

Just catch the error from `ItemTree::sort()` and clear the
secondary-sort and fallback-sort prefs so that sorting works on the next
attempt.

Replacement for a8ed30ce803a9c5b3b5c04d50f6bd6b94faeee86

https://groups.google.com/g/zotero-dev/c/kc0-C6-SA74/m/bhHniGceAQAJ

dstillman committed:
Fix test failure after 26a5e2900a

mrtcode committed:
Include quotation marks for highlights by default, but allow to disable
Fixes #2697

mrtcode committed:
Don't include quotes when {{highlight}} is placed in a blockquote
Fixes #2697

mrtcode committed:
Prevent delayed context pane expansion and pdf-reader rerender/resize when switching from library tab

AbeJellinek committed:
Prevent deleting My Publications

Previously showed a blank confirmation dialog and logged an error.

mrtcode committed:
Use Quick Copy (Markdown) for direct plain-text copying from note-editor
Fixes #2505

dstillman committed:
Add warning in toolbar for 'dev' channel builds

dstillman committed:
Apply Big Sur styling when macOS version comes through correctly

Follow-up to zotero/zotero-standalone-build@f7a506a32 -- now that some
people are somehow getting an accurate macOS version number, we need to
apply the Big Sur CSS to all higher versions.

https://forums.zotero.org/discussion/98406/zotero-will-not-open-mac-os-monterey

dstillman committed:
Update pdf-reader submodule

dstillman committed:
Increase Mendeley API timeout to 60 seconds

To see if we can fix the 30-second timeout here:

https://forums.zotero.org/discussion/comment/412750/#Comment_412750

dstillman committed:
Disable sort tests from 26a5e2900a for now

Until I can figure out the race-condition breakage they trigger later in
the file

dstillman committed:
Update locales from Transifex

dstillman committed:
Update submodules

dstillman committed:
Update translate submodule

dstillman committed:
Update note-editor submodule

dstillman committed:
Update version

dstillman committed:
Localize "Export PDF…" menu option

Fixes #2710

adomasven committed:
Update SingleFile submodule

adomasven committed:
Changes for updated SingleFile submodule

adomasven committed:
Fixes some assets not being saved with SingleFile snapshots

Part of zotero/zotero-connectors#394

When saving an item (e.g. on arxiv) with a site translator, the
translator may add a custom snapshot attachment URL which does not match
the translation URL (the URL in the browser where translation was
initiated) in which case Zotero handles snapshot saving. Zotero loads
the page in a background browser. By default images are not loaded in
this browser. SingleFile due to CORS restrictions has to refetch all
resources that are not from the same domain or subdomain, but uses
pre-loaded resources (images) from the same domain. Before this change
any images that were from the same domain as the saved website were not
saved in the snapshot. This commit fixes that.

mrtcode committed:
Fix regression that prevents showing new notes in contextPane notes list

dstillman committed:
Update global schema

dstillman committed:
Update version

mrtcode committed:
Fix regression that prevents showing the last selected tab on startup
Fixes #2723

mrtcode committed:
Freeze note schema version to 8
We can do it because it was only used to create a note from annotations.

No need to update schema version in Zotero client, unless using new
features when creating a note from annotations.

adomasven committed:
Do not load-timeout on browser PDF download during content load

Closes #2726

AbeJellinek committed:
Duplicates Merge: Preserve embedded annotations (#2728)

mrtcode committed:
Allow to scroll tabs with mouse wheel
Fixes #2724

mrtcode committed:
Select existing unloaded tab instead of opening a new one
Fixes #2740

AbeJellinek committed:
Item box: Prevent tabbing to hidden fields

This one's probably been around for a while! Reproduce by creating an
item with a fieldMode = 1 creator, tabbing past the creator, and then
shift-tabbing back to it. Your cursor will end up in the invisible first
name field and further shift-tabs can't move it past.

Cherry-picked from fx102: 080ada78eea782176e5e9e4f10b17be9d1f4ea08

AbeJellinek committed:
Item box: Fix fieldMode resetting after editor hidden twice

hideEditor() called switchCreatorMode() too early, setting the fieldMode
attribute on the soon-to-be-discarded textbox instead of the label
replacing it. Then, showing the editor a second time would carry over an
empty fieldMode attribute from the label to the new textbox. Hiding that
editor would update the creator in the item to fieldMode = 0 and trigger
a save.

Moving the switchCreatorMode() call does the trick, and the flex
settings changes still work fine when made there.

dstillman committed:
Don't preselect Act for multiple sources in classic citation dialog

"Multiple Sources…" wasn't properly handled by 629892b0b8

https://forums.zotero.org/discussion/98877/act-instead-of-page
https://twitter.com/drezvera99/status/1554521940684984320

AbeJellinek committed:
Advanced Search: Make sure items are loaded (#2743)

AbeJellinek committed:
Add test for 87455d

dstillman committed:
Automatically substitute `event-title` for `event` in styles

Stopgap until styles are updated in citation-style-language/styles#6151

dstillman committed:
Reset `Zotero.locale` to 'en-US' after cite test

dstillman committed:
Improve test for `event-title` substitution

And add comments

Follow-up to 20d0f103f

mrtcode committed:
Fix regression preventing opening PDF reader window when the tab is open

mrtcode committed:
Fix PDF reader selector popup that was never opening

adomasven committed:
Update SingleFile config for current version

They got updated upstream and the backwards-compat code
does not run for the way we execute singlefile

See https://github.com/gildas-lormeau/SingleFile/issues/1021
Addresses https://github.com/zotero/zotero-connectors/issues/398

adomasven committed:
Prevent singlefile config from overwriting the Zotero.SingleFile namespace

AbeJellinek committed:
Add Book -> Book Section and Book Section -> Book to context menu (#2755)

dstillman committed:
Update locales from Transifex and merge new English strings

dstillman committed:
Update global schema for locales

dstillman committed:
Update translators

dstillman committed:
Fix copying of standalone attachments between libraries

Regression in f684984b07 (maybe a race condition)

dstillman committed:
Don't relate book sections to each other when creating from book

Fixes #2757

dstillman committed:
Fix importing of standalone attachments

Maybe has been broken for years?

https://forums.zotero.org/discussion/99020/warning-data-loss-when-exporting-collections

dstillman committed:
Fix importing of standalone attachments without URLs

And allow `importSnapshotFromFile()` without `parentItemID` or
`libraryID`, which is more consistent with other attachment functions

Follow-up to 318e4852e9

https://forums.zotero.org/discussion/comment/414802/#Comment_414802

dstillman committed:
Update pdf-reader submodule

dstillman committed:
Update locales from Transifex

dstillman committed:
Update translators

dstillman committed:
Match parent attachments for annotation tags

Expose annotation tags in tag selector and match parent attachments when
filtering/searching

This also fixes searching for annotation text or comments when using
Everything quick search.

This is temporary until we display annotations in the items list
directly.

dstillman committed:
Update version

dstillman committed:
Update pdf-reader submodule

dstillman committed:
Fix and disable item-tree test for emptying trash

20c6fe6737d caused this to start failing, but only because the test was
testing something too specific. The change in that commit caused more
rows to be left behind when emptying the trash (for reasons I should
probably look into), but the trash wasn't being emptied properly before,
which #2606 should fix. This test should be restored as part of that PR.

dstillman committed:
Update localized string for pdf-reader

AbeJellinek committed:
Scaffold: Non-async detectWeb and async scrape by default

We don't usually want an async detectWeb, since HTTP requests should only be
used there in very exceptional cases. We do usually want an async scrape (and we
were already - mistakenly - awaiting it).

mrtcode committed:
Fix regression for broken "Snow in Page"
Fixes #2764
Regression from 7ec54fd

mrtcode committed:
Fix navigation to annotation when opening an unloaded PDF reader tab
Regression from 142e3b09

tnajdek committed:
Fix importing Citavi annotations in some cases

tnajdek committed:
Make citavi annotations importer more resilient

mrtcode committed:
EditorInstance.createNoteFromAnnotations(): Handle multiple attachments

AbeJellinek committed:
Empty Trash: Visually remove deleted items, refresh icon (#2606)

There were a few problems causing the incorrect behavior:
1. Rows were being removed only if they had no non-deleted children, which
   wasn't the right check. We want to remove all rows with no *deleted*
   children.
2. Children of the removed rows weren't being removed with them.
3. We weren't invalidating the tree (which _removeRows() doesn't do).

Also:

* Erase trashed annotation after getAnnotations() test

Because ItemTree#notify() doesn't yet correctly handle refresh events on
parent items that are themselves children (three-level nesting: item ->
attachment -> annotation), this test was causing a failure in
itemTreeTest.js.

dstillman committed:
Fixed filed items with annotations appearing in Unfiled Items

Fixes #2771

Regression from 20c6fe67

dstillman committed:
Include proper test for b373291c02 for #2771

dstillman committed:
Update pdf-reader submodule

dstillman committed:
Update translators

dstillman committed:
Update version

dstillman committed:
Don't repeatedly upload relations

Fixes #2772

adomasven committed:
Fix SingleFile snapshot saving stalling in Firefox Connector

Update SingleFile submodule

dstillman committed:
Use items-list sort order for report from collection or saved search

https://forums.zotero.org/discussion/99284/exporting-a-report-sorted-by-first-date

Regression from HTML tree

dstillman committed:
Add Zotero.Item::topLevelItem and Zotero.Items.getTopLevel(items)

dstillman committed:
Rename Zotero.Items::keepParents() to keepTopLevel()

So that it works for more than two levels of items

Also fix a bug where the parent item could be returned more than once if
multiple child items were selected.

mrtcode committed:
Update note-editor submodule and strings for math and table insertion

dstillman committed:
Merge new English strings

AbeJellinek committed:
duplicateAndConvertSelectedItem: Don't copy abstracts (#2799)

dstillman committed:
Fix HTTP.request() 5xx handling with errorDelayMax=0

This was causing delay-less infinite loops for OPTIONS requests to Box,
which are returning 502 errors.

dstillman committed:
Update note-editor submodule

dstillman committed:
Update pdf-reader submodule

mrtcode committed:
Add missing `noteEditor.mathBlock` string

mrtcode committed:
Update pdf-reader submodule

mrtcode committed:
Allow inserting images over context menu
Fixes #2759

dstillman committed:
Add `collectionID` option to EditorInstance.createNoteFromAnnotations()

For adding a standalone note to the current collection

dstillman committed:
Add "Create Note from Annotations" option to items list context menu

To allow creation of a standalone note with annotations from all the
selected top-level items and/or attachments. Annotations will be sorted
by the order of the items in the items list.

"Add Note from Annotations" remains when a single regular item or one or
more attachments under a single regular item are selected.

dstillman committed:
Merge new English strings

mrtcode committed:
Add context menu options to copy/save image from pdf-reader/note-editor
Fixes #2664

dstillman committed:
Merge new English strings

AbeJellinek committed:
Hide "Find Available PDF" for feed items (#2807)

mrtcode committed:
Adjust PDF reader and context pane UI based on font size pref
Fixes #2809

tnajdek committed:
Map annotation colors colors on import #2819 (#2822)

mrtcode committed:
Convert Mendeley colors to equivalent Zotero colors on PDF open
Fixes #2820

mrtcode committed:
Fix PDF reader sometimes showing UTC time and sometimes local

dstillman committed:
Update translate submodule

mrtcode committed:
Fix `focusContextPane` error when pressing tab in standalone PDF window
Fixes #2823

tnajdek committed:
Make creator buttons accessible via keyboard #999

dstillman committed:
Update ESLint config

dstillman committed:
Additional tweaks for creator button keyboard access (#2826)

- Open creator type menu on ArrowUp as well
- Fix outline on Windows/Linux
- Don't show focus ring on hover, since we already show a background

dstillman committed:
Merge pull request #2826 from tnajdek/999-keyboard-nav-creators

Make creator buttons accessible via keyboard

dstillman committed:
Update pdf-reader and pdf-worker submodules

dstillman committed:
Update locales from Transifex and merge new English strings

dstillman committed:
Update package-lock.json

dstillman committed:
Add Number as items list column option

https://forums.zotero.org/discussion/99844/add-columns-to-desktop-app-report-number

dstillman committed:
Update note-editor submodule

dstillman committed:
Update translators and CSL locales

dstillman committed:
Update locales from Transifex

mrtcode committed:
Add split view for PDF reader (#2832)

mrtcode committed:
Import external annotations before trying to add a note from annotations

mrtcode committed:
Prevent error when creating a note from annotations that have ink

mrtcode committed:
Update note-editor submodule

dstillman committed:
Update version

dstillman committed:
Update locales from Transifex and merge new English strings

dstillman committed:
Swap "Split Vertically" and "Split Horizontally" in PDF reader

dstillman committed:
Remove stray "aaa" in "Split Horizontally" in PDF window

From 12b1897478

dstillman committed:
Remove default OpenURL gateway

https://www.worldcat.org/registry/gateway is no longer supported -- see
https://forums.zotero.org/discussion/comment/418373/#Comment_418373 for
details

If no resolver is selected, show an error on "Library Lookup" with a
button to open the prefs.

dstillman committed:
Merge new English strings

dstillman committed:
Update locales from Transifex

dstillman committed:
Update version

dstillman committed:
Update pdf-reader submodule

dstillman committed:
Add 15 and 16 as note font size options in menu

https://forums.zotero.org/discussion/100079/notebook-fonts

mrtcode committed:
Remove PDF view rotation options

adomasven committed:
Fix item list row text not read when using non-contiguous kb selection

Closes #2830

mrtcode committed:
Re-render annotation images when rotating a page
Fixes #2824

mrtcode committed:
Don't allow pdf-reader to set annotation author name
Fixes the issue when your own created annotation becomes grayish and indicates that it was modified by another user.

mrtcode committed:
Allow importing annotations in group libraries

jaeschrich committed:
Exposed toolbar functionality via the menubar

dstillman commented:
This seems to include an outdated version of the code from the other PR? Unless this depends on that PR, in which case you should just push the new commit to that one, this PR should only include the relevant commits.

dstillman commented:
> I can refactor this to avoid duplication in one of two ways: (1) by creating an XBL binding that encapsulates the shared menu items, or (2) by creating a JavaScript function that sets up both the menubar popup and the toolbarbutton popup identically.

(2) for sure. XBL is discontinued in fx102.

For sync, I would add a Sync Now option to the Tools menu (like Firefox) and not worry about cancel or status for the moment.

dstillman commented:
Location tweaks:

- Add Attachment and Add Item by Identifier should go in File, below a separator below New*
- Locate can go in File, since it's about the selected item (i.e., the "file") — View is really about the UI, not viewing things externally
- Sync Now can go at the top of Tools, in its own section

Other things:

- At least on macOS, Locate isn't being fully populated based on the current item selection. Crossref Lookup and Google Scholar Search appear, but not Library Lookup or the View options above. If a note is selected, it says "No items selected", so it's at least partially populating it.
- In a read-only library, the New Note menuitem itself isn't disabled, like New Item. (Only the child items are disabled.) Same for Add Attachment and Add Item by Identifier. (The former has disabled child items, but the latter actually opens the panel.)

jaeschrich committed:
moved menuitems based on feedback. added "Sync Now" button to Tools menu

jaeschrich committed:
disabled New Note, Add Attachment, and Add Item by Identifier for read-only libraries

jaeschrich commented:
Hey Dan,
I moved the menuitems as requested and added a "Sync Now" button. I also wrote some code to keep the menu items in sync with the toolbarbuttons. I fixed the bug with the menuitems not getting disabled properly in read-only libraries. 

I can't reproduce the bug with the "locate" menu not being populated. I tried it on both windows and linux, and it works on both. It's possible that the new code to dynamically populate the menuitems will make it work on Mac; otherwise, I'm not sure what else to try.
